### DISCUSSION TITLE
cue -> css

### DISCUSSION THREAD

**Discussion by mbarkhau** [cue-lang/cue#1172] [Ideas]
After introducing a colleague to cue recently, he remarked that it reminds him of sass. Now I'm wondering if css might be an interesting output format for cue. If anybody tried something like this I would love to hear about your experience, or have a look at your work.

**Comment by verdverm** [comment]
Generally, for `CUE -> <file>`, where it is not a structured data file like Yaml or JSON, you can use the `text/template` package, a.k.a. text interpolation. With the `cue` cli, you can wrap this process in a script, or use the tooling layer to write the generated templates.

I've written https://github.com/hofstadter-io/hof for large scale output generation from CUE. Think having a single-source of truth designs to generate in multiple languages across your tech stack. It also uses diff3 so you can work in both the design space and the output. A core feature is that the schemas and templates are separate from the tool, so they can be written independently in their own repositories and shared as modules.

You can find examples of both techniques on this page: https://cuetorials.com/first-steps/generate-all-the-things/

**Reply by verdverm** [reply]
Some related discussions:

- https://github.com/cue-lang/cue/discussions/1027
- https://github.com/cue-lang/cue/discussions/1038
- https://github.com/cue-lang/cue/discussions/482

**Comment by myitcv** [comment]
@mbarkhau can you describe in a bit more detail what you/your colleague are imagining here, perhaps by way of example?

**Reply by mbarkhau** [reply]
The idea is to write style rules in cue with very shallow selectors and completely avoid the cascading aspect of css. Instead you have rule templates with defaults, placeholders, embedding etc. which you later make concrete.

In other words, you don't need to think about all the other styles in the universe that might override your layout as cue will not emit anything if there are conflicting values. This is all very vague and wishful thinking at the moment.

**Reply by myitcv** [reply]
Understood, thanks. I find the best way to explore these sorts of things is to try and write out, by hand, examples. Examples of the CUE you would want/expect to write, the equivalent CSS, then stand back and see what it looks like. If it still makes sense at that point, I would either look to try and answer whether an encoding between the two can be supported or if plain code generation in one direction only suffices/is the only option.
